contracts/src/CrossChainApplications/NativeTokenBridge/ERC20TokenSource.sol

```diff
@@ -0,0 +1,226 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {IWarpMessenger} from "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import {IERC20TokenSource} from "./IERC20TokenSource.sol";
+import {ITokenSource} from "./ITokenSource.sol";
+import {ITeleporterMessenger, TeleporterMessageInput, TeleporterFeeInfo} from "../../Teleporter/ITeleporterMessenger.sol";
+import {ITeleporterReceiver} from "../../Teleporter/ITeleporterReceiver.sol";
+import {SafeERC20TransferFrom} from "../../Teleporter/SafeERC20TransferFrom.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract ERC20TokenSource is
+    ITeleporterReceiver,
+    IERC20TokenSource,
+    ITokenSource,
+    ReentrancyGuard
+{
+    // The address where the burned transaction fees are credited.
+    address public constant BURNED_TX_FEES_ADDRESS =
+        0x0100000000000000000000000000000000000000;
+    uint256 public constant MINT_NATIVE_TOKENS_REQUIRED_GAS = 100_000;
+    // Used to keep track of tokens burned through transactions on the destination chain. They can
+    // be reported to this contract to burn an equivalent number of tokens on this chain.
+    uint256 public destinationChainBurnedBalance = 0;
```

Comment:
nit: think we're avoiding `destinationChain` in fvaor of `destinationBlockchain`, could just be `destinationBurnedBalance` too